### PR TITLE
Azure Cloud Provider should support Service annotations that allow for ip-tag control over the public ips created for LoadBalancer Services

### LINKED ISSUE #94113: Azure Cloud Provider should support Service annotations that allow for ip-tag control over the public ips created for LoadBalancer Services
**What would you like to be added**:

When the Azure Cloud Provider creates public-ips and binds them to LoadBalancer endpoints, it should allow for customer control of the ip-tags passed along to public-ip creation requests.

**Why is this needed**:

There are optional features in Azure that allow customers to control ip allocations in nuanced ways, based on the ip-tags that are set during public ip creation.  Today, if customers would like to use these, they need to statically pre-create their own public-ip objects, grant kubernetes clusters access to them, and then reference them by name in their CRDs (via azure-pip-name).   This is brittle and requires coordinated life-cycle management.   Supporting these controls directly allows for true, Kubernetes-based dynamic public-ip creation for a broader range of scenarios.

### PR DESCRIPTION
**Pull request by MarcPow**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
See Issue #94113 

**Which issue(s) this PR fixes**:
Fixes #94113

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
Yes.

```release-note
Support [service.beta.kubernetes.io/azure-pip-ip-tags] annotations to allow customers to specify ip-tags to influence public-ip creation in Azure [Tag1=Value1, Tag2=Value2, etc.]
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go [hunk 1]
```diff
@@ -1482,27 +1565,54 @@ func (az *Cloud) reconcilePublicIP(clusterName string, service *v1.Service, lbNa
...
+		if
+		// If we don't actually want a load balancer,
+		!wantLb ||
+			// If the load balancer is internal, and thus doesn't require public exposure
+			isInternal ||
+			// If the name of this public ip does not match the desired name,
+			(pipName != desiredPipName) ||
+			// If the service annotations have specified required ip tags, but thy do not match the iptags of the existing instance
+			(serviceIPTags != nil && (pipPropertiesFormat == nil /*unit test*/ ||
+				(pipPropertiesFormat != nil && areIPTagsEqualOrderInsensitive((*pipPropertiesFormat).IPTags, serviceIPTags)))) {
```
Comment (nilo19) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go:1598]: I think this line doesn't reflect the comment above. Here it means "if the ip tags setting in the service match the ones in the existing pip".

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go [hunk 2]
```diff
@@ -586,6 +591,83 @@ func (az *Cloud) ensurePublicIPExists(service *v1.Service, pipName string, domai
...
+
+	if ipTags2 == nil {
+		ipTags2 = &[]network.IPTag{}
+	}
+
+	return reflect.DeepEqual(ipTags1, ipTags2)
+}
+
+func convertIPTagMapToSlice(ipTagMap map[string]string) *[]network.IPTag {
+
```
Comment (feiskyer) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go:650]: nit: remove empty line

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go [hunk 3]
```diff
@@ -1482,27 +1565,54 @@ func (az *Cloud) reconcilePublicIP(clusterName string, service *v1.Service, lbNa
...
+		if
+		// If we don't actually want a load balancer,
+		!wantLb ||
+			// If the load balancer is internal, and thus doesn't require public exposure
+			isInternal ||
+			// If the name of this public ip does not match the desired name,
+			(pipName != desiredPipName) ||
+			// If the service annotations have specified required ip tags, but thy do not match the iptags of the existing instance
+			(serviceIPTags != nil && (pipPropertiesFormat == nil /*unit test*/ ||
```
Comment (feiskyer) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go:1597]: I think we should use `len(serviceIPTags) != 0` here, as `make(map[string]string)` is not nil

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go (hunk 1 above)
Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go:1598]: yea, should add `!` here

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go [hunk 4]
```diff
@@ -1482,27 +1565,54 @@ func (az *Cloud) reconcilePublicIP(clusterName string, service *v1.Service, lbNa
...
+
+		// Now, let's perform additional analysis.
+		// Specifically, we should decide if the discovered public ips should be released/deleted.
+
+		// If the service does not own the ip, we cannot delete it under any circumstances
+		if !serviceOwnsPublicIP(&pip, clusterName, serviceName) {
+			continue
+		}
+
+		if
```
Comment (feiskyer) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go:1589]: the conditions here are a little complicated, could we convert it to a function?

**Review by feiskyer — CHANGES_REQUESTED** [review]

**Review by feiskyer — APPROVED** [review]
thanks for fixing the comments

/lgtm
/approve

**Comment by feiskyer** [comment]
Pending master branch reopen after v1.19.0 release

**Comment by neolit123** [comment]
please see this PR merge policy change.
https://groups.google.com/g/kubernetes-dev/c/YXGBa6pxLzo

the milestone should be applied later in the cycle.
for Q you can ask in #sig-release.

thanks

/milestone clear

**Comment by MarcPow** [comment]
@feiskyer - FYI: Squashed commits to prepare for merge when 1.20 opens.

**Comment by MarcPow** [comment]
/test pull-kubernetes-e2e-azure-disk-windows
ref: #94269

**Pull request merged**